### ISSUE TITLE
Multi Dex Android Apps fail on repeating builds

### ISSUE THREAD

**Issue by mn-martin** [NativeScript/nativescript-cli#2686]
$ tns --version
2.5.3

In `~/app/App_Resources/app.gradle` I added the following config as suggested by [Eddy Verbruggen](https://github.com/EddyVerbruggen/nativescript-plugin-firebase/blob/ae482bfdfcb4f590a4cd8e2426fcf8bb21c4b65f/README.md#dexindexoverflowexception):

```
android {
  defaultConfig {
    multiDexEnabled true
  }
} 
```

This introduced the following Exception when running builds for android:

```
Exception in thread "main" java.lang.IllegalArgumentException: id.support.v7.widget.Toolbar_OnMenuItemClickListener conflict: ~/platforms/android/build/intermediates/transforms/dex/F0F1F2F3F4F5F6F7F8/debug/folders/1000/1f/main/classes4.dex and ~/platforms/android/build/intermediates/transforms/dex/F0F1F2F3F4F5F6F7F8/debug/folders/1000/1f/main/classes6.dex
```

I found out that if i keep running builds even thought I'd always get that Exception there would be more .dex files added with increasing index in that directory. Then I manually removed the directory, ran a build and everything worked. Running one more build again gave me that Exception until I removed the directory yet again.

My current solution to this is removing that directory before each build using npm scripts:

```
  "scripts": {
    "clear-dex-files": "rm -r platforms/android/build/intermediates/transforms/dex/ || true",
    "run-android": "npm run clear-dex-files && tns run android --no-watch --clean",
    "run-ios": "tns run ios --no-watch --clean",
    "reinit-android": "tns platform remove android && tns run android --no-watch --clean",
    "reinit-ios": "tns platform remove ios && tns run ios --no-watch --clean"
  },
```

I gues the CLI is not working correctly on multi dex android apps yet? Can you guys fix this in the CLI or is there a better work around using hooks? (I've tried using a hook but failed because I could not find any documentation about hooks that run before the actual build)

**Comment by Pip3r4o** [comment]
Hey @mn-martin 

Thanks for reporting this issue.

There isn't any specific logic in the CLI, nor in the android runtime build scripts, that clean up the build artifacts accordingly when the application is configured to build with multidex.

You could manually run `./platforms/android/gradlew clean` to ensure that the build directory will be purged.

As a temporary solution you could also make an after-prepare hook to do the clean up (or just delete the `build` directory `rm -rf platforms/android/build`) on every prepare. The process of creating your own hooks is described in the following readme -> https://github.com/NativeScript/nativescript-hook

**Comment by mn-martin** [comment]
This seems to be fixed (at least for me) in 2.5.4 (tns --version)
I no longer have to execute the clear-dex-files script.

**Comment by Pip3r4o** [comment]
Hi @mn-martin 

There are no changes in 2.5.4 related to android builds. I would guess that gradle got confused between builds and didn't clean up properly.

Thanks for the follow up, closing the issue as resolved.
